### DISCUSSION TITLE
Whitespaces in docstrings and Sphinx

### DISCUSSION THREAD

**Discussion by b-goldsmith** [unitaryfund/mitiq#2567] [General]
I am working on the calibration tables for the API docs that currently look like:
![image](https://github.com/user-attachments/assets/c990cc60-3955-4318-8902-e39d0ed05baa)

Is anybody familiar with a different/better way to have Sphinx retain whitespaces in a docstring? 

I was able to get it working using:
![image](https://github.com/user-attachments/assets/454af2be-67c2-48de-8498-a798e60010f0)

This was the only way I was able to get it to work. The table has to be under `Example:` and the first line must start with `>>>`. This puts a `<pre>` tag around everything indented under `Example`.

Looking up more about how `>>>` is used, it turns out that it is part of [doctest](https://docs.python.org/3/library/doctest.html).  While I don't think anybody is directly running `calibrator.py` (with or without the -v), it is a bit of a hacky-use in this situation. I did think about including the command lines that generate the table, but it gets a bit tricky. The `log_results_flat()` and `log_results_cartesian()` are applied to the results of running a calibrator which requires an executor and if we want the specific output in the docs, we would need custom settings. It does not make sense to me to put all of that in a docstring that's already pretty big due to the tables. 

Anybody have any other ideas?

**Comment by purva-thakre** [comment]
You could use the `list table` directive in sphinx if the goal is to retain docstring formatting of a table.

**Comment by b-goldsmith** [comment]
@purva-thakre Thanks for the suggestion. I tried playing around with `list table` and `table` directives. While I don't think those will apply to this, **I did find that I can get the table to print correctly using `::`** (as long as it is not in the first line of the docstring.) So, by doing the following it works:
![image](https://github.com/user-attachments/assets/097809d9-5e69-4b13-aea1-aebd7c36b99f)

![image](https://github.com/user-attachments/assets/15b5091b-9e1a-45b1-ad23-16c1aaa3b155)


If `flat form::` is at the end of the first line (even if the entire line is moved down), the formatting does not work since Sphinx sees it as a title of a section (notice how the sentence is bold.)
![image](https://github.com/user-attachments/assets/28609595-8a1b-4cb3-8cc9-9e1bb9c2fb33)

![image](https://github.com/user-attachments/assets/3adc27c6-5fcf-4ffa-aa67-19742706cf67)

**Comment by purva-thakre** [comment]
@bdg221 Sphinx expects the table to be of a particular form. 

https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#tables

The example docstring table in the function definition is copy-pasted from the function output directly where the formatting of columns/rows does not match with how sphinx expects you to format a table. 

There are several options on how you can format a table for sphinx:

- https://myst-parser.readthedocs.io/en/latest/syntax/tables.html#markdown-syntax
- https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#tables

I just tried the following and it works as expected:

![image](https://github.com/user-attachments/assets/2fdc8f5c-ba30-4cf9-b8f8-9939ab289aac) ![image](https://github.com/user-attachments/assets/5e0659fe-464a-4aea-bb5a-9d980d3c7cfe)

I am not too familiar with `flatform::`. I would say if that works, then go with the option that doesn't require a lot of work.

**Reply by b-goldsmith** [reply]
The table in the docstring is an example of the output table, so it has to look exactly the same. This is more about needing Sphinx to retain whitespaces than print a table in a specific format. 

![image](https://github.com/user-attachments/assets/220f5108-c489-4ed1-821e-88d4bc55e72f)

**Reply by purva-thakre** [reply]
I understand what you are getting at. What I am trying to say is, with the way the table is added in the docstring, you can't use things like `:align: left` or `:widths: 15` to set the whitespace. You might have to change the formatting to able to use the thing you want.

**Reply by purva-thakre** [reply]
@bdg221 What if you embed a screenshot of the table? This way you don't have to worry about formatting issues. 

It also solves the problem of the table appearing exactly as that of the function output. 

https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#images

ETA: As I said in the previous comment, since `flatform::` works as well, I think that might be the easiest option as well.

**Comment by purva-thakre** [comment]
Closing as this appears to be fixed by #2568. 

Feel free to reopen if we need to continue this discussion.
